### PR TITLE
ARROW-1663: [Java] use consistent name for null and not-null in FixedSizeLis…

### PR DESCRIPTION
**Pull request by yufeldman**
…t, add backward compatibility while deserializing schema that was generated before this JIRA checkin

### CODE REVIEW

File: java/vector/src/main/java/org/apache/arrow/vector/ZeroVector.java [hunk 1]
```diff
@@ -109,7 +109,7 @@ public void clear() {
 
   @Override
   public Field getField() {
-    return new Field(name, FieldType.nullable(new Null()), null);
+    return new Field(NAME, FieldType.nullable(new Null()), null);
```
Comment (jacques-n) [inline java/vector/src/main/java/org/apache/arrow/vector/ZeroVector.java]: why not just change the name to $data$?

Comment (yufeldman) [reply java/vector/src/main/java/org/apache/arrow/vector/ZeroVector.java]: It would be ideal I think, as in this case we won't need changes to ListVector and FixedSizeListVector at all.
Just may be part for backward compatibility.
I need to verify that "default" is not used for any other purpose though.

**Comment by wesm** [comment]
What is the status of this patch? Can you rebase?

File: java/vector/src/main/java/org/apache/arrow/vector/ZeroVector.java [hunk 2]
```diff
@@ -109,7 +111,7 @@ public void clear() {
 
   @Override
   public Field getField() {
-    return new Field(name, FieldType.nullable(new Null()), null);
+    return new Field(DATA_VECTOR_NAME, FieldType.nullable(new Null()), null);
```
Comment (BryanCutler) [inline java/vector/src/main/java/org/apache/arrow/vector/ZeroVector.java]: I don't think we should fix the name here to `BaseRepeatedValueVector.DATA_VECTOR_NAME`.  The default should just be a generic name, I think what was there before was fine.

File: java/vector/src/main/java/org/apache/arrow/vector/complex/FixedSizeListVector.java [hunk 3]
```diff
@@ -54,6 +53,8 @@
 import org.apache.arrow.vector.util.SchemaChangeRuntimeException;
 import org.apache.arrow.vector.util.TransferPair;
 
+import io.netty.buffer.ArrowBuf;
+
```
Comment (BryanCutler) [inline java/vector/src/main/java/org/apache/arrow/vector/complex/FixedSizeListVector.java]: undo this change

**Review by BryanCutler — CHANGES_REQUESTED** [review]
Thanks for the PR @yufeldman!  In general, I think we should keep the changes simple and just repeat what was done in ARROW-1347 for FixedSizeListVector.  Otherwise, I think we should hold off until refactoring in ARROW-1447 is complete to discuss more.

What is your use case for making schema backward compatible?  I believe we are not making that guarantee right now.

Could you also add the tag [Java] to the PR title?

File: java/vector/src/main/java/org/apache/arrow/vector/ZeroVector.java (hunk 2 above)
Comment (jacques-n) [reply java/vector/src/main/java/org/apache/arrow/vector/ZeroVector.java]: This causes weird problems if you're trying to merge different schemas since it suggests that the inner vector changed if we moved from a List<null> to a List<int>, thus the change. The idea that the vector name is different before and after we populate the list doesn't really make any sense. Inner vectors should always have the same name.

**Comment by jacques-n** [comment]
We aren't making the guarantee at the Arrow level that schema is backwards compatible. However, we have a bunch of Dremio OSS and Enterprise users that are relying on this code and the inconsistency here has been causing us problems (when dealing with data that is initially untyped lists but then is later found to have a type). I don't see much issue to adding the backwards compatibility code and it should help others who might have hit a problem. What are your concerns @BryanCutler?

**Comment by BryanCutler** [comment]
@jacques-n , there's no problem with adding backwards compatibility - I just wasn't sure where the request was coming from so I was trying to get some context. I'm fine with this, but I think some of it could be simplified a little.

File: java/vector/src/main/java/org/apache/arrow/vector/ZeroVector.java (hunk 2 above)
Comment (BryanCutler) [reply java/vector/src/main/java/org/apache/arrow/vector/ZeroVector.java]: That's fine, but my concern was that is `ZeroVector` only meant to ever be an inner data vector?  Or would a user ever want to make a ZeroVector and not name it `DATA_VECTOR_NAME`? If not then I think we should just hardcode the name "$data$" like you mentioned before. Otherwise maybe we can do something like allow setting a name in the constructor, and create the instance with the name "$data$".

File: java/vector/src/main/java/org/apache/arrow/vector/types/pojo/Field.java [hunk 4]
```diff
@@ -121,7 +124,18 @@ public static Field convertField(org.apache.arrow.flatbuf.Field field) {
...
+        && childField.getName().equals(ZeroVector.NAME)) {
+        Field modifiedField = new Field(DATA_VECTOR_NAME, childField.isNullable(), childField.getType(),
+          childField.getDictionary(),
+          childField.getChildren(),
+          childField.getTypeLayout(),
+          childField.getMetadata());
+        childrenBuilder.add(modifiedField);
+        continue;
+      }
+      childrenBuilder.add(childField);
```
Comment (BryanCutler) [inline java/vector/src/main/java/org/apache/arrow/vector/types/pojo/Field.java]: Could you put this block in a separate method with a name that makes it obvious what it is changing and why?

Comment (yufeldman) [reply java/vector/src/main/java/org/apache/arrow/vector/types/pojo/Field.java]: sure

File: java/vector/src/main/java/org/apache/arrow/vector/complex/FixedSizeListVector.java (hunk 3 above)
Comment (yufeldman) [reply java/vector/src/main/java/org/apache/arrow/vector/complex/FixedSizeListVector.java]: sure

File: java/vector/src/main/java/org/apache/arrow/vector/types/pojo/Field.java (hunk 4 above)
Comment (yufeldman) [reply java/vector/src/main/java/org/apache/arrow/vector/types/pojo/Field.java]: Any other comments?

File: java/vector/src/main/java/org/apache/arrow/vector/ZeroVector.java [hunk 5]
```diff
@@ -39,7 +41,7 @@
 public class ZeroVector implements FieldVector {
   public final static ZeroVector INSTANCE = new ZeroVector();
 
-  private final String name = "[DEFAULT]";
+  public static final String NAME = "[DEFAULT]";
```
Comment (BryanCutler) [inline java/vector/src/main/java/org/apache/arrow/vector/ZeroVector.java]: If this name is just used for the conversion, can we move it there?

File: java/vector/src/test/java/org/apache/arrow/vector/TestListVector.java [hunk 6]
```diff
@@ -633,11 +633,11 @@ public void testGetBufferAddress() throws Exception {
...
     try (ListVector listVector = ListVector.empty("sourceVector", allocator)) {
       String emptyListStr = listVector.getField().toString();
-      assertTrue(emptyListStr.contains(ListVector.DATA_VECTOR_NAME));
+      Assert.assertTrue(emptyListStr.contains(ListVector.DATA_VECTOR_NAME));
 
-      listVector.addOrGetVector(FieldType.nullable(MinorType.INT.getType()));
+      listVector.addOrGetVector(FieldType.nullable(Types.MinorType.INT.getType()));
       String emptyVectorStr = listVector.getField().toString();
-      assertTrue(emptyVectorStr.contains(ListVector.DATA_VECTOR_NAME));
+      Assert.assertTrue(emptyVectorStr.contains(ListVector.DATA_VECTOR_NAME));
```
Comment (BryanCutler) [inline java/vector/src/test/java/org/apache/arrow/vector/TestListVector.java]: The rest of the tests in this file match the way it was previously.  If you're going to change it here then I would prefer making the others consistent as well.

File: java/vector/src/test/java/org/apache/arrow/vector/pojo/TestConvert.java [hunk 7]
```diff
@@ -48,6 +55,73 @@
...
+    "    } ],\n" +
+    "    \"typeLayout\" : {\n" +
+    "      \"vectors\" : [ {\n" +
+    "        \"type\" : \"VALIDITY\",\n" +
+    "        \"typeBitWidth\" : 1\n" +
+    "      } ]\n" +
+    "    },\n" +
+    "    \"name\" : \"a\"\n" +
+    "  } ]\n" +
+    "}\n";
```
Comment (BryanCutler) [inline java/vector/src/test/java/org/apache/arrow/vector/pojo/TestConvert.java]: This is a large block that's very hard to read just to check the name of a field.  Is there a more compact way you can test this?

File: java/vector/src/test/java/org/apache/arrow/vector/TestListVector.java (hunk 6 above)
Comment (yufeldman) [reply java/vector/src/test/java/org/apache/arrow/vector/TestListVector.java]: done

File: java/vector/src/main/java/org/apache/arrow/vector/ZeroVector.java (hunk 5 above)
Comment (yufeldman) [reply java/vector/src/main/java/org/apache/arrow/vector/ZeroVector.java]: done

File: java/vector/src/test/java/org/apache/arrow/vector/pojo/TestConvert.java (hunk 7 above)
Comment (yufeldman) [reply java/vector/src/test/java/org/apache/arrow/vector/pojo/TestConvert.java]: done

**Review by yufeldman — COMMENTED** [review]
Addressed review comments

File: java/vector/src/main/java/org/apache/arrow/vector/types/pojo/Field.java [hunk 8]
```diff
@@ -47,6 +49,8 @@
 
 public class Field {
 
+  public static final String ZEROVECTOR_OLD_NAME = "[DEFAULT]";
```
Comment (BryanCutler) [inline java/vector/src/main/java/org/apache/arrow/vector/types/pojo/Field.java]: Just move this directly into `mutateOriginalNameIfNeeded` and same with the tests, no need to make this a static variable in Field.

File: java/vector/src/test/java/org/apache/arrow/vector/pojo/TestConvert.java [hunk 9]
```diff
@@ -65,6 +71,43 @@ public void complex() {
...
+    try (BufferAllocator allocator = new RootAllocator(Long.MAX_VALUE);
+         ListVector writeVector = ListVector.empty("list", allocator)) {
+      Field listVectorField = writeVector.getField();
+      childrenBuilder.add(listVectorField);
+    }
+    try (BufferAllocator allocator = new RootAllocator(Long.MAX_VALUE);
+         FixedSizeListVector writeVector = FixedSizeListVector.empty("fixedlist", 5, allocator)) {
+      Field listVectorField = writeVector.getField();
+      childrenBuilder.add(listVectorField);
+    }
```
Comment (BryanCutler) [inline java/vector/src/test/java/org/apache/arrow/vector/pojo/TestConvert.java]: What are these 2 try blocks testing for?

File: java/vector/src/test/java/org/apache/arrow/vector/pojo/TestConvert.java [hunk 10]
```diff
@@ -65,6 +71,43 @@ public void complex() {
...
+    ImmutableList.Builder<Field> parentBuilder = ImmutableList.builder();
+    parentBuilder.add(initialField);
+    FlatBufferBuilder builder = new FlatBufferBuilder();
+    builder.finish(initialField.getField(builder));
+    org.apache.arrow.flatbuf.Field flatBufField = org.apache.arrow.flatbuf.Field.getRootAsField(builder.dataBuffer());
+    Field finalField = Field.convertField(flatBufField);
+    assertEquals(initialField, finalField);
+    assertTrue(!finalField.toString().contains(Field.ZEROVECTOR_OLD_NAME));
+
+
```
Comment (BryanCutler) [inline java/vector/src/test/java/org/apache/arrow/vector/pojo/TestConvert.java]: nit: remove 1 blank line

File: java/vector/src/test/java/org/apache/arrow/vector/pojo/TestConvert.java [hunk 11]
```diff
@@ -65,6 +71,43 @@ public void complex() {
...
+    Field initialField = new Field("a", FieldType.nullable(Struct.INSTANCE), childrenBuilder.build());
+    ImmutableList.Builder<Field> parentBuilder = ImmutableList.builder();
+    parentBuilder.add(initialField);
+    FlatBufferBuilder builder = new FlatBufferBuilder();
+    builder.finish(initialField.getField(builder));
+    org.apache.arrow.flatbuf.Field flatBufField = org.apache.arrow.flatbuf.Field.getRootAsField(builder.dataBuffer());
+    Field finalField = Field.convertField(flatBufField);
+    assertEquals(initialField, finalField);
+    assertTrue(!finalField.toString().contains(Field.ZEROVECTOR_OLD_NAME));
```
Comment (BryanCutler) [inline java/vector/src/test/java/org/apache/arrow/vector/pojo/TestConvert.java]: change to `assertFalse`

File: java/vector/src/test/java/org/apache/arrow/vector/pojo/TestConvert.java [hunk 12]
```diff
@@ -65,6 +71,43 @@ public void complex() {
...
+    Schema initialSchema = new Schema(parentBuilder.build());
+    String jsonSchema = initialSchema.toJson();
+    String modifiedSchema = jsonSchema.replace("$data$", "[DEFAULT]");
+
+    Schema tempSchema = Schema.fromJSON(modifiedSchema);
+    FlatBufferBuilder schemaBuilder = new FlatBufferBuilder();
+    org.apache.arrow.vector.types.pojo.Schema schema = new org.apache.arrow.vector.types.pojo.Schema(tempSchema.getFields());
+    schemaBuilder.finish(schema.getSchema(schemaBuilder));
+    Schema finalSchema = Schema.deserialize(ByteBuffer.wrap(schemaBuilder.sizedByteArray()));
+    assertTrue(!finalSchema.toString().contains(Field.ZEROVECTOR_OLD_NAME));
```
Comment (BryanCutler) [inline java/vector/src/test/java/org/apache/arrow/vector/pojo/TestConvert.java]: assertFalse instead

File: java/vector/src/test/java/org/apache/arrow/vector/pojo/TestConvert.java [hunk 13]
```diff
@@ -65,6 +71,43 @@ public void complex() {
...
+    }
+    try (BufferAllocator allocator = new RootAllocator(Long.MAX_VALUE);
+         FixedSizeListVector writeVector = FixedSizeListVector.empty("fixedlist", 5, allocator)) {
+      Field listVectorField = writeVector.getField();
+      childrenBuilder.add(listVectorField);
+    }
+
+    Field initialField = new Field("a", FieldType.nullable(Struct.INSTANCE), childrenBuilder.build());
+    ImmutableList.Builder<Field> parentBuilder = ImmutableList.builder();
+    parentBuilder.add(initialField);
```
Comment (BryanCutler) [inline java/vector/src/test/java/org/apache/arrow/vector/pojo/TestConvert.java]: is this `parentBuilder` used?

File: java/vector/src/test/java/org/apache/arrow/vector/pojo/TestConvert.java (hunk 9 above)
Comment (yufeldman) [reply java/vector/src/test/java/org/apache/arrow/vector/pojo/TestConvert.java]: yes

File: java/vector/src/test/java/org/apache/arrow/vector/pojo/TestConvert.java (hunk 13 above)
Comment (yufeldman) [reply java/vector/src/test/java/org/apache/arrow/vector/pojo/TestConvert.java]: further down to get initial schema

File: java/vector/src/test/java/org/apache/arrow/vector/pojo/TestConvert.java (hunk 10 above)
Comment (yufeldman) [reply java/vector/src/test/java/org/apache/arrow/vector/pojo/TestConvert.java]: ok

File: java/vector/src/test/java/org/apache/arrow/vector/pojo/TestConvert.java (hunk 12 above)
Comment (yufeldman) [reply java/vector/src/test/java/org/apache/arrow/vector/pojo/TestConvert.java]: ok

File: java/vector/src/test/java/org/apache/arrow/vector/pojo/TestConvert.java (hunk 11 above)
Comment (yufeldman) [reply java/vector/src/test/java/org/apache/arrow/vector/pojo/TestConvert.java]: ok

File: java/vector/src/main/java/org/apache/arrow/vector/types/pojo/Field.java (hunk 8 above)
Comment (yufeldman) [reply java/vector/src/main/java/org/apache/arrow/vector/types/pojo/Field.java]: mutateOriginalNameIfNeeded may be invoked many times. I understand that compiler will optimize it, but I am not sure it makes sense to multiply "[DEFAULT]". May be place to keep this const is not a great one, I would prefer to keep it in ZeroVector - at least it has some association with it.

**Comment by siddharthteotia** [comment]
Changes look good to me. I am assuming we have pre-tested the patch with Dremio.
Please prefix the PR title with [JAVA].

**Comment by wesm** [comment]
+1
